### PR TITLE
WIP: Support richmenu api

### PR DESCRIPTION
**Pull request by musou1500**
**This PR is WORKING IN PROGRESS**

* [x] GET /richmenu/{richMenuId}
* [x] POST /richmenu
* [x] DELETE /richmenu/{richMenuId}
* [x] GET /user/{userId}/richmenu
* [x] POST /user/{userId}/richmenu/{richMenuId}
* [x] DELETE /user/{userId}/richmenu/{richMenuId}
* [x] GET /richmenu/{richMenuId}/content
* [x] POST /richmenu/{richMenuId}/content
* [x] GET /richmenu/list

### CODE REVIEW

**Comment by musou1500** [comment]
This PR will resolves #33

**Comment by musou1500** [comment]
I have conciderations below about `POST /richmenu/{richMenuId}/content`.

1. What type is disirable for request body?
I am considering to use `Buffer` or `Stream`.
https://github.com/axios/axios#request-config

2. It's OK to make developer to explictly specify Content-Type? 

I think signature of the function will looks like below.
```typescript
  public uploadRichMenuContent(data: Buffer | Stream, contentType: string);
```

**Comment by utatti** [comment]
@musou1500 

1. I think supporting both `Buffer` and `Stream` is a great idea.
2. Having `contentType` is okay, but we can consider making it optional `contentType?: string` and detecting mime type from buffer.
    + https://github.com/sindresorhus/image-type

**Comment by musou1500** [comment]
@utatti thank you. I'll implement content-type detection by using [file-type](https://www.npmjs.com/package/file-type) and [file-type-stream](https://www.npmjs.com/package/file-type-stream)
[image-type](https://github.com/sindresorhus/image-type) use it internally.

**Comment by musou1500** [comment]
I implemented it.

File: lib/http.ts [hunk 1]
```diff
@@ -64,3 +66,32 @@ export function post(url: string, headers: any, data?: any): Promise<any> {
...
+  headers: any,
+  data: Buffer | Readable,
+  contentType?: string,
+): Promise<Readable> {
+  return new Promise(resolve => {
+    return resolve(contentType ? contentType : detectContentType(data));
+  }).then((contentType: string) => {
+    headers["Content-Type"] = contentType;
+    headers["User-Agent"] = userAgent;
+    const responseType = "stream";
```
Comment (utatti) [inline lib/http.ts]: It's upload, not download, and response type don't need to be `stream`. Just use `checkJSON` and return JS object.

File: test/client.spec.ts [hunk 2]
```diff
@@ -143,4 +168,90 @@ describe("client", () => {
...
+
+  it("getUserRichMenuIds", () => {
+    return client.getUserRichMenuIds("test_user_id").then((res: any) => {
+      equal(res.headers.authorization, "Bearer test_channel_access_token");
+      equal(res.path, "/user/test_user_id/richmenu");
+      equal(res.method, "GET");
+    });
+  });
+
+  it("linkRichMenuWithUser", () => {
```
Comment (utatti) [inline test/client.spec.ts]: Don't forget to rename test after renaming method.

File: test/client.spec.ts [hunk 3]
```diff
@@ -143,4 +168,90 @@ describe("client", () => {
...
+    return client
+      .linkRichMenuWithUser("test_user_id", "test_rich_menu_id")
+      .then((res: any) => {
+        equal(res.headers.authorization, "Bearer test_channel_access_token");
+        equal(res.path, "/user/test_user_id/richmenu/test_rich_menu_id");
+        equal(res.method, "POST");
+      });
+  });
+
+  it("unlinkRichMenuWithUser", () => {
```
Comment (utatti) [inline test/client.spec.ts]: Don't forget to rename test after renaming method.

File: test/http.spec.ts [hunk 4]
```diff
@@ -75,6 +83,38 @@ describe("http", () => {
...
+    const buffer = readFileSync(filepath);
+    return postBinary(`${TEST_URL}/post`, testHeaders, buffer)
+      .then(s => getStreamData(s))
+      .then((data: string) => {
+        const res = JSON.parse(data);
+        equal(res.method, "POST");
+        equal(res.path, "/post");
+        equal(res.headers["test-header-key"], testHeaders["test-header-key"]);
+        equal(res.headers["user-agent"], `${pkg.name}/${pkg.version}`);
+        equal(res.headers["content-type"], "image/png");
```
Comment (utatti) [inline test/http.spec.ts]: Can you add an assert to check if the submitted file is the same as origin? A `multipart/form-data` parser need to be added to test server.

File: test/http.spec.ts [hunk 5]
```diff
@@ -75,6 +83,38 @@ describe("http", () => {
...
+      });
+  });
+
```
Comment (utatti) [inline test/http.spec.ts]: Please add a test case to upload a pipe with `postBinary`.

File: lib/types.ts [hunk 6]
```diff
@@ -270,3 +270,21 @@ export type TemplateDatetimePickerAction = {
...
   min?: string;
 };
+
+export type Size = {
+  width: number;
+  height: number;
+};
+
+export type RichMenuArea = {
+  bounds: ImageMapArea;
```
Comment (utatti) [inline lib/types.ts]: I think it's better renaming `ImageMapArea` into `Area`, as it's now used not only in `ImageMap`.

File: lib/types.ts [hunk 7]
```diff
@@ -270,3 +270,21 @@ export type TemplateDatetimePickerAction = {
...
+  action: TemplateAction<any>;
```
Comment (utatti) [inline lib/types.ts]: Can you also rename `TemplateAction` to `Action`?

File: lib/types.ts [hunk 8]
```diff
@@ -270,3 +270,21 @@ export type TemplateDatetimePickerAction = {
...
+};
+
+export type RichMenu = {
+  size: Size;
+  selected: boolean;
+  name: string;
+  chatBarText: string;
+  areas: RichMenuArea[];
```
Comment (utatti) [inline lib/types.ts]: `RichMenuArea` is only used here and needn't be a separate type.

``` typescript
{
  ...
  areas: Array<{ bounds: Area, action: Action }>;
}
```

File: lib/client.ts [hunk 9]
```diff
@@ -102,10 +103,60 @@ export default class Client {
...
+  }
+
+  public unlinkRichMenuWithUser(
+    userId: string,
+    richMenuId: string,
+  ): Promise<any> {
+    return this.delete(URL.userRichMenu(userId, richMenuId));
+  }
+
+  public getRichMenuContent(richMenuId: string): Promise<any> {
```
[the remark this answers is not in the document]
Comment (utatti) [reply lib/client.ts]: I've discussed with other bot dev members, and the method name is decided to be `getRichMenuImage`.

File: lib/client.ts [hunk 10]
```diff
@@ -102,10 +103,60 @@ export default class Client {
...
+    return this.stream(URL.richMenuContent(richMenuId));
+  }
+
+  public uploadRichMenuContent(
```
Comment (utatti) [inline lib/client.ts]: I've discussed with other bot dev members, and the method name is decided to be `uploadRichMenuImage`.

File: test/client.spec.ts [hunk 11]
```diff
@@ -143,4 +168,90 @@ describe("client", () => {
...
+    return client
+      .unlinkRichMenuWithUser("test_user_id", "test_rich_menu_id")
+      .then((res: any) => {
+        equal(res.headers.authorization, "Bearer test_channel_access_token");
+        equal(res.path, "/user/test_user_id/richmenu/test_rich_menu_id");
+        equal(res.method, "DELETE");
+      });
+  });
+
+  it("uploadRichMenuContent", () => {
```
Comment (utatti) [inline test/client.spec.ts]: Don't forget to rename test after renaming method.

File: test/client.spec.ts [hunk 12]
```diff
@@ -143,4 +168,90 @@ describe("client", () => {
...
+      .then(s => getStreamData(s))
+      .then((data: string) => {
+        const res = JSON.parse(data);
+        equal(res.headers.authorization, "Bearer test_channel_access_token");
+        equal(res.path, "/richmenu/test_rich_menu_id/content");
+        equal(res.method, "POST");
+      });
+  });
+
+  it("getRichMenuContent", () => {
```
Comment (utatti) [inline test/client.spec.ts]: Don't forget to rename test after renaming method.

File: lib/http.ts (hunk 1 above)
Comment (musou1500) [reply lib/http.ts]: I understand it. 
This API returns empty response when succeeded to upload, but returns JSON when failed.
so I think I can't consistently treat response as JSON.
What is the ideal implementation? It's ok to branch by status code?

**Comment by musou1500** [comment]
@utatti thank you for review.
 It's not related to those comments, but which should I use `NodeJS.ReadableStream` or `import { Readable } from "stream"` ?

File: lib/http.ts (hunk 1 above)
Comment (utatti) [reply lib/http.ts]: Ah, you are right.

I guess non-200 responses are handled by `wrapError`.

For empty response,  Only file upload and DELETEs use empt response according to the official doc, so I think we can just skip `checkJSON` and return `res.data` as it is. Can you please also remove `checkJSON` from DELETE?

**Comment by utatti** [comment]
`import { Readable } from "stream"` seems better. Can you also replace all `NodeJS.ReadableStream` occurrences with `Readable`?

File: test/http.spec.ts (hunk 4 above)
Comment (musou1500) [reply test/http.spec.ts]: as you know, the api requires `image/png` or `image/jpeg` for `Content-Type`.
https://developers.line.me/ja/docs/messaging-api/reference/#anchor-4f3b10c3de51f6728025dc81cc8f1efc63fd6235
so `postBinary` doesn't uses `multipart/form-data`.
I considering a way such as pipe request body to writable file stream.
Do you have better idea to achieve it?

**Comment by musou1500** [comment]
@utatti ok. I'll do it.

File: test/http.spec.ts (hunk 4 above)
Comment (utatti) [reply test/http.spec.ts]: Ah, you're right. We can make this into a separate work. I will investigate on this after this PR is merged, so please do not care of this comment 😅

File: test/http.spec.ts (hunk 5 above)
Comment (utatti) [reply test/http.spec.ts]: This too.

File: lib/http.ts (hunk 1 above)
Comment (musou1500) [reply lib/http.ts]: thank you. ok.

**Comment by musou1500** [comment]
@utatti 
I began to fix, so I'll report the progress.

the following tasks remain.
* add test to check to whether data is read from stream correctly after detectContentType.

and I have a question about the following line in `client.ts`.
```typescript
      // TODO: change return type to appropriate one
      public linkRichMenuToUser(userId: string, richMenuId: string): Promise<any> {
        return this.post(URL.userRichMenu(userId, richMenuId));
      }
```
`http.post` function always throws error when invalid JSON was returned.

but the API returns empty response. How should I do?
https://developers.line.me/ja/docs/messaging-api/reference/#anchor-51d49588bbbedf60c3ea20a8c553f04680e6f9ca

File: lib/http.ts [hunk 13]
```diff
@@ -64,3 +63,30 @@ export function post(url: string, headers: any, data?: any): Promise<any> {
...
+  data: Buffer | Readable,
+  contentType?: string,
+): Promise<any> {
+  return new Promise(resolve => {
+    return resolve(contentType ? contentType : detectContentType(data));
+  }).then((contentType: string) => {
+    headers["Content-Type"] = contentType;
+    headers["User-Agent"] = userAgent;
+    return axios
+      .post(url, data, { headers, responseType: "text" })
```
Comment (utatti) [inline lib/http.ts]: I think `responseType` is not needed?

File: lib/http.ts [hunk 14]
```diff
@@ -64,3 +63,30 @@ export function post(url: string, headers: any, data?: any): Promise<any> {
...
+      .then(res => res.data)
+      .catch(wrapError);
+  });
+}
+
+export function del(url: string, headers: any): Promise<any> {
+  headers["User-Agent"] = userAgent;
+
+  return axios
+    .delete(url, { headers, responseType: "text" })
```
Comment (utatti) [inline lib/http.ts]: I think `responseType` is not needed?

File: lib/types.ts [hunk 15]
```diff
@@ -237,10 +237,10 @@ export type TemplateImageCarousel = {
 
 export type TemplateImageColumn = {
   imageUrl: string;
-  action: TemplateAction<{ label?: string }>;
+  action: Action<{ label?: string }>;
 };
 
-export type TemplateAction<Label> =
+export type Action<Label> =
   | TemplatePostbackAction & Label
```
Comment (utatti) [inline lib/types.ts]: Please remove `Template-` prefixes of summed types too, e.g. `TemplatePostbackAction` → `PostbackAction`.

File: lib/util.ts [hunk 16]
```diff
@@ -1,3 +1,20 @@
+import { Readable } from "stream";
+import * as fileType from "file-type";
+const { default: fileTypeStream } = require("file-type-stream");
```
Comment (utatti) [inline lib/util.ts]: Destructuring doesn't work in Node.js 4. Please use a plain `require("file-type-stream").default`.

**Comment by utatti** [comment]
I've left some more comments. Please take a look at them.

Also, here are my responds for your questions.

1. About test cases (including stream body test), I will work on them after this PR is merged. Please do not worry about them.
2. Currently Axios takes care of JSON parsing, and I think it's not really meaningful to keep `checkJSON`. Can you please follow the instructions below and remove `checkJSON` in `http.ts`?
    + `lib/http.ts`: remove `checkJSON` and make `get()` and `post()` just `.then(res => res.data)`
    + `test/http.spec.ts`: remove a test case concerning `JSONParseError`
    + `docs/pages/guide/client.md`: remove a line mentioning `JSONParseError`

**Comment by musou1500** [comment]
@utatti sorry if you already know it. axios seems not to throws exception when fail to parse. Is it OK to delete `checkJSON`?
https://github.com/axios/axios/blob/188334439f9698c8ddd58697f75e7f0b745c9acc/lib/defaults.js#L60

**Comment by utatti** [comment]
That was my intention, but now I think it's better to keep them on APIs where JSON response should be used. I apologies for the confusion. How about moving `checkJSON` into `client.ts` and checking it only when return type is actually JSON, not `any`?

**Comment by musou1500** [comment]
@utatti ok. I'll move `checkJSON` into `client.ts`. thank you.

**Comment by musou1500** [comment]
@utatti 
done.
 * remove `Template` prefixes of summed types
 * fix return type of `getRichMenuList`
 * remove `Template` prefixes of summed types
 * fix return type of `getRichMenuList`
 * fixed to not to use destructuring assignment

I'm sorry about `detectContentType`, it is my mistake.
I need to add `next` arg to `detectContentType`.
I think fixed version of the function will be below.

```typescript
export function detectContentType(
  data: Readable,
  next: Writable,
): Promise<string>;

export function detectContentType(data: Buffer): Promise<string>;

export function detectContentType(
  data: Buffer | Readable,
  next?: Writable,
): Promise<string> {
  return new Promise(resolve => {
    const handleFileTypeResult = (result: any) =>
      resolve(result ? result.mime : null);

    if (data instanceof Buffer) {
      handleFileTypeResult(fileType(data));
    } else {
      data.pipe(fileTypeStream(handleFileTypeResult)).pipe(next);
    }
  });
}
```

**Comment by utatti** [comment]
Yes, I thought about that too. We can separate it to another issue.

File: lib/client.ts [hunk 17]
```diff
@@ -22,7 +23,7 @@ export default class Client {
     return this.post(URL.push, {
       messages: toArray(messages),
       to,
-    });
+    }) /*.then(this.checkJSON)*/;
```
Comment (utatti) [inline lib/client.ts]: We needn't care about response here (anyway the response will be empty, thus return type is `any`), so just remove the check.

File: lib/client.ts [hunk 18]
```diff
@@ -32,7 +33,7 @@ export default class Client {
     return this.post(URL.reply, {
       messages: toArray(messages),
       replyToken,
-    });
+    }).then(this.checkJSON);
```
Comment (utatti) [inline lib/client.ts]: We needn't care about response here (anyway the response will be empty, thus return type is `any`), so just remove the check.

File: lib/client.ts [hunk 19]
```diff
@@ -42,52 +43,58 @@ export default class Client {
     return this.post(URL.multicast, {
       messages: toArray(messages),
       to,
-    });
+    }).then(this.checkJSON);
```
Comment (utatti) [inline lib/client.ts]: We needn't care about response here (anyway the response will be empty, thus return type is `any`), so just remove the check.

File: lib/client.ts [hunk 20]
```diff
@@ -96,32 +103,31 @@ export default class Client {
   }
 
   public leaveGroup(groupId: string): Promise<any> {
-    return this.post(URL.leaveGroup(groupId));
+    return this.post(URL.leaveGroup(groupId)).then(this.checkJSON);
```
Comment (utatti) [inline lib/client.ts]: We needn't care about response here (anyway the response will be empty, thus return type is `any`), so just remove the check.

File: lib/client.ts [hunk 21]
```diff
@@ -96,32 +103,31 @@ export default class Client {
...
   }
 
   public leaveRoom(roomId: string): Promise<any> {
-    return this.post(URL.leaveRoom(roomId));
+    return this.post(URL.leaveRoom(roomId)).then(this.checkJSON);
```
Comment (utatti) [inline lib/client.ts]: We needn't care about response here (anyway the response will be empty, thus return type is `any`), so just remove the check.

**Comment by musou1500** [comment]
@utatti I'm really thankful to you. I removed unnecessary `checkJSON`.

**Comment by utatti** [comment]
Thanks again for the fantastic works. I'll look into the test stuff.

**Pull request merged**